File: tools/compare_subset0_json.exs
#!/usr/bin/env elixir
Code.require_file("cort_tooling.ex", __DIR__)

defmodule CompareSubset0Json do
  @blocking_classifications MapSet.new(["required", "required-but-error-text-flexible"])
  @boolean_semantic_fields ["pointer_identity_same"]
  @note_fields [
    "type_id",
    "retain_count_before",
    "retain_count_after_retain",
    "retain_count_after_release",
    "retain_callback_count",
    "release_callback_count"
  ]

  def main(argv) do
    options =
      CortTooling.parse_args!(argv,
        fx_json: :string,
        mx_json: :string,
        fx_label: :string,
        mx_label: :string,
        output: :string
      )

    fx_path = CortTooling.required_flag!(options, :fx_json)
    mx_path = CortTooling.required_flag!(options, :mx_json)
    fx_label = CortTooling.string_flag(options, :fx_label) || fx_path
    mx_label = CortTooling.string_flag(options, :mx_label) || mx_path
    output_path = CortTooling.string_flag(options, :output)

    fx_results = load_results!(fx_path)
    mx_results = load_results!(mx_path)
    {report, blockers, _warnings} = render_report(fx_results, mx_results, fx_label, mx_label)

    CortTooling.render_or_write!(report, output_path)
    System.halt(if blockers == 0, do: 0, else: 1)
  rescue
    error in [ArgumentError, File.Error, ErlangError, RuntimeError] ->
      CortTooling.usage_and_exit!(Exception.message(error))
  end

  defp load_results!(path) do
    data = CortTooling.read_json!(path)
    results = Map.get(data, "results")

    unless is_list(results) do
      raise ArgumentError, "#{path} does not contain a top-level results array"
    end

    Enum.reduce(results, %{}, fn entry, acc ->
      name = Map.get(entry, "name")

      cond do
        not (is_binary(name) and name != "") ->
          raise ArgumentError, "#{path} contains a result without a valid name"

        Map.has_key?(acc, name) ->
          raise ArgumentError, "#{path} contains duplicate result name #{inspect(name)}"

        true ->
          Map.put(acc, name, entry)
      end
    end)
  end

  defp render_report(fx_results, mx_results, fx_label, mx_label) do
    all_names =
      Map.keys(fx_results)
      |> Enum.concat(Map.keys(mx_results))
      |> Enum.uniq()
      |> Enum.sort()

    initial_lines = [
      "# Subset 0 FX vs MX Comparison",
      "",
      "- FX input: `#{fx_label}`",
      "- MX input: `#{mx_label}`",
      ""
    ]

    if all_names == [] do
      {Enum.join(initial_lines ++ ["No cases were found in either input.", ""], "\n"), 0, 0}
    else
      {rows, blockers, warnings} =
        Enum.reduce(all_names, {[], 0, 0}, fn name, {acc, blocker_count, warning_count} ->
          {line, delta_blockers, delta_warnings} =
            compare_case(name, Map.get(fx_results, name), Map.get(mx_results, name))

          {[line | acc], blocker_count + delta_blockers, warning_count + delta_warnings}
        end)

      verdict =
        if blockers == 0 do
          "- verdict: no blocking mismatches found in the shared comparison surface"
        else
          "- verdict: blocking mismatches found; do not treat FX as semantically aligned for the affected required rows"
        end

      report_lines =
        initial_lines ++
          [
            "| Case | Status | Notes |",
            "| --- | --- | --- |"
          ] ++
          Enum.reverse(rows) ++
          [
            "",
            "- blockers: #{blockers}",
            "- warnings: #{warnings}",
            verdict,
            ""
          ]

      {Enum.join(report_lines, "\n"), blockers, warnings}
    end
  end

  defp compare_case(name, nil, mx_case) do
    missing_case_line(name, "FX", "MX", mx_case)
  end

  defp compare_case(name, fx_case, nil) do
    missing_case_line(name, "MX", "FX", fx_case)
  end

  defp compare_case(name, fx_case, mx_case) do
    fx_class = Map.get(fx_case, "classification")
    mx_class = Map.get(mx_case, "classification")
    blocking_case = blocking?(fx_class) or blocking?(mx_class)

    {notes, status, blockers, warnings} =
      {[], "match", 0, 0}
      |> maybe_add_classification_note(fx_class, mx_class)
      |> maybe_add_success_note(blocking_case, Map.get(fx_case, "success"), Map.get(mx_case, "success"))
      |> maybe_add_boolean_field_notes(blocking_case, fx_case, mx_case)
      |> add_diagnostic_field_notes(fx_case, mx_case)

    final_notes =
      if status == "match" and notes == [] do
        ["semantically aligned on shared blocking fields"]
      else
        notes
      end

    {"| `#{name}` | #{status} | #{Enum.join(final_notes, "; ")} |", blockers, warnings}
  end

  defp missing_case_line(name, missing_side, present_side, present_case) do
    classification =
      if is_map(present_case) do
        Map.get(present_case, "classification")
      else
        nil
      end

    notes = "missing on #{missing_side}; present on #{present_side} as #{classification || "unknown"}"

    if blocking?(classification) do
      {"| `#{name}` | BLOCKER | #{notes} |", 1, 0}
    else
      {"| `#{name}` | warning | #{notes} |", 0, 1}
    end
  end

  defp maybe_add_classification_note({notes, status, blockers, warnings}, fx_class, mx_class) do
    if fx_class != mx_class do
      {
        notes ++ ["classification differs: FX=#{inspect(fx_class)} MX=#{inspect(mx_class)}"],
        if(status == "BLOCKER", do: status, else: "warning"),
        blockers,
        warnings + if(status == "BLOCKER", do: 0, else: 1)
      }
    else
      {notes, status, blockers, warnings}
    end
  end

  defp maybe_add_success_note({notes, status, blockers, warnings}, blocking_case, fx_success, mx_success) do
    if fx_success != mx_success do
      note = "success differs: FX=#{inspect(fx_success)} MX=#{inspect(mx_success)}"

      if blocking_case do
        {
          notes ++ [note],
          "BLOCKER",
          blockers + if(status == "BLOCKER", do: 0, else: 1),
          warnings
        }
      else
        {
          notes ++ [note],
          if(status == "BLOCKER", do: status, else: "warning"),
          blockers,
          warnings + if(status in ["warning", "BLOCKER"], do: 0, else: 1)
        }
      end
    else
      {notes, status, blockers, warnings}
    end
  end

  defp maybe_add_boolean_field_notes({notes, status, blockers, warnings}, blocking_case, fx_case, mx_case) do
    Enum.reduce(@boolean_semantic_fields, {notes, status, blockers, warnings}, fn field, acc ->
      fx_value = Map.get(fx_case, field)
      mx_value = Map.get(mx_case, field)

      if is_nil(fx_value) or is_nil(mx_value) or fx_value == mx_value do
        acc
      else
        {acc_notes, acc_status, acc_blockers, acc_warnings} = acc
        note = "#{field} differs: FX=#{inspect(fx_value)} MX=#{inspect(mx_value)}"

        if blocking_case do
          {
            acc_notes ++ [note],
            "BLOCKER",
            acc_blockers + if(acc_status == "BLOCKER", do: 0, else: 1),
            acc_warnings
          }
        else
          {
            acc_notes ++ [note],
            if(acc_status == "BLOCKER", do: acc_status, else: "warning"),
            acc_blockers,
            acc_warnings + if(acc_status in ["warning", "BLOCKER"], do: 0, else: 1)
          }
        end
      end
    end)
  end

  defp add_diagnostic_field_notes({notes, status, blockers, warnings}, fx_case, mx_case) do
    updated_notes =
      Enum.reduce(@note_fields, notes, fn field, acc ->
        fx_value = Map.get(fx_case, field)
        mx_value = Map.get(mx_case, field)

        if is_nil(fx_value) or is_nil(mx_value) or fx_value == mx_value do
          acc
        else
          acc ++ ["#{field} differs: FX=#{inspect(fx_value)} MX=#{inspect(mx_value)} (diagnostic)"]
        end
      end)

    {updated_notes, status, blockers, warnings}
  end

  defp blocking?(classification), do: MapSet.member?(@blocking_classifications, classification)
end

CompareSubset0Json.main(System.argv())

File: tools/compare_subset0_json.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path


BLOCKING_CLASSIFICATIONS = {"required", "required-but-error-text-flexible"}
BOOLEAN_SEMANTIC_FIELDS = {"pointer_identity_same"}
NOTE_FIELDS = {
    "type_id",
    "retain_count_before",
    "retain_count_after_retain",
    "retain_count_after_release",
    "retain_callback_count",
    "release_callback_count",
}


def load_results(path: Path) -> dict[str, dict]:
    data = json.loads(path.read_text(encoding="utf-8"))
    results = data.get("results")
    if not isinstance(results, list):
        raise ValueError(f"{path} does not contain a top-level results array")

    by_name: dict[str, dict] = {}
    for entry in results:
        name = entry.get("name")
        if not isinstance(name, str) or not name:
            raise ValueError(f"{path} contains a result without a valid name")
        if name in by_name:
            raise ValueError(f"{path} contains duplicate result name {name!r}")
        by_name[name] = entry
    return by_name


def is_blocking(classification: str | None) -> bool:
    return classification in BLOCKING_CLASSIFICATIONS


def render_report(
    fx_results: dict[str, dict],
    mx_results: dict[str, dict],
    fx_label: str,
    mx_label: str,
) -> tuple[str, int, int]:
    blockers = 0
    warnings = 0
    lines: list[str] = []
    lines.append("# Subset 0 FX vs MX Comparison")
    lines.append("")
    lines.append(f"- FX input: `{fx_label}`")
    lines.append(f"- MX input: `{mx_label}`")
    lines.append("")

    all_names = sorted(set(fx_results) | set(mx_results))
    if not all_names:
        lines.append("No cases were found in either input.")
        return "\n".join(lines) + "\n", blockers, warnings

    lines.append("| Case | Status | Notes |")
    lines.append("| --- | --- | --- |")

    for name in all_names:
        fx_case = fx_results.get(name)
        mx_case = mx_results.get(name)
        notes: list[str] = []
        status = "match"

        if fx_case is None or mx_case is None:
            present = fx_case if fx_case is not None else mx_case
            present_side = "FX" if fx_case is not None else "MX"
            missing_side = "MX" if fx_case is not None else "FX"
            classification = present.get("classification") if isinstance(present, dict) else None
            notes.append(f"missing on {missing_side}; present on {present_side} as {classification or 'unknown'}")
            if is_blocking(classification):
                status = "BLOCKER"
                blockers += 1
            else:
                status = "warning"
                warnings += 1
            lines.append(f"| `{name}` | {status} | {'; '.join(notes)} |")
            continue

        fx_class = fx_case.get("classification")
        mx_class = mx_case.get("classification")
        blocking_case = is_blocking(fx_class) or is_blocking(mx_class)

        if fx_class != mx_class:
            notes.append(f"classification differs: FX={fx_class!r} MX={mx_class!r}")
            if status != "BLOCKER":
                status = "warning"
                warnings += 1

        fx_success = fx_case.get("success")
        mx_success = mx_case.get("success")
        if fx_success != mx_success:
            notes.append(f"success differs: FX={fx_success!r} MX={mx_success!r}")
            if blocking_case:
                status = "BLOCKER"
                blockers += 1
            elif status != "warning":
                status = "warning"
                warnings += 1

        for field in sorted(BOOLEAN_SEMANTIC_FIELDS):
            fx_value = fx_case.get(field)
            mx_value = mx_case.get(field)
            if fx_value is None or mx_value is None:
                continue
            if fx_value != mx_value:
                notes.append(f"{field} differs: FX={fx_value!r} MX={mx_value!r}")
                if blocking_case:
                    if status != "BLOCKER":
                        status = "BLOCKER"
                        blockers += 1
                elif status != "warning":
                    status = "warning"
                    warnings += 1

        for field in sorted(NOTE_FIELDS):
            fx_value = fx_case.get(field)
            mx_value = mx_case.get(field)
            if fx_value is None or mx_value is None:
                continue
            if fx_value != mx_value:
                notes.append(f"{field} differs: FX={fx_value!r} MX={mx_value!r} (diagnostic)")

        if status == "match" and not notes:
            notes.append("semantically aligned on shared blocking fields")

        lines.append(f"| `{name}` | {status} | {'; '.join(notes)} |")

    lines.append("")
    lines.append(f"- blockers: {blockers}")
    lines.append(f"- warnings: {warnings}")
    if blockers == 0:
        lines.append("- verdict: no blocking mismatches found in the shared comparison surface")
    else:
        lines.append("- verdict: blocking mismatches found; do not treat FX as semantically aligned for the affected required rows")
    lines.append("")
    return "\n".join(lines), blockers, warnings


def main() -> int:
    parser = argparse.ArgumentParser(description="Compare FX and MX Subset 0 JSON results.")
    parser.add_argument("--fx-json", required=True, help="Path to the FX JSON artifact")
    parser.add_argument("--mx-json", required=True, help="Path to the MX JSON artifact")
    parser.add_argument("--fx-label", help="Optional label to render instead of the FX JSON path")
    parser.add_argument("--mx-label", help="Optional label to render instead of the MX JSON path")
    parser.add_argument("--output", help="Optional path for the markdown report")
    args = parser.parse_args()

    fx_path = Path(args.fx_json)
    mx_path = Path(args.mx_json)

    fx_results = load_results(fx_path)
    mx_results = load_results(mx_path)
    report, blockers, _warnings = render_report(
        fx_results,
        mx_results,
        args.fx_label or str(fx_path),
        args.mx_label or str(mx_path),
    )

    if args.output:
        output_path = Path(args.output)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        output_path.write_text(report, encoding="utf-8")
    else:
        sys.stdout.write(report)

    return 1 if blockers else 0


if __name__ == "__main__":
    sys.exit(main())
